### DISCUSSION TITLE
Question: Save with custom columns

### DISCUSSION THREAD

**Discussion by janpeterka** [excid3/noticed#317] [Q&A] [answered]
Hi, this may be stupid question, I'm quite new to Rails, so some things are not obvious to me.

My `notifications` table has some additional columns that I need to save data to. Main is `content_message` column, which I need to be the same as `:content` value. I'm unable to set this, tried:
- `.with(announcement:, content: 'created', content_message: 'created')`
- in `before_delivery`
- in Notification model `before_validate`

Thanks for help!

**Comment by excid3** [comment, marked as answer]
See https://github.com/excid3/noticed/blob/main/docs/delivery_methods/database.md and https://github.com/excid3/noticed/blob/a970f6831a5b5419e8253813d899142f61175fc6/lib/noticed/delivery_methods/database.rb#L6

```ruby
class MyNotification do
  deliver_by :database, format: :custom_attributes

  def custom_attributes
    notification.params.slice(:announcement, :content, :content_message).merge(
      type: notification.class.name,
      params: notification.params
    )
  end
end
```

**Reply by janpeterka** [reply]
Sorry, missed that. Thank you!

**Reply by excid3** [reply]
We'll be updating the docs to make this more obvious soon. 👍 

Version 2 will have a smoother way of defining database attributes.

**Reply by artur1313** [reply]
@excid3 any update? I do not see anything in docs about how to save extra attributes to database models

**Reply by excid3** [reply]
It's supported and they're just ActiveRecord models so you can set attributes like you normally would.

```ruby
CommentNotifier.new(attribute: value, params: {}).deliver(User.last)
```

**Reply by artur1313** [reply]
thx, but what if i want save extra data to `Noticed::Notification` instance? In my case i want save body of notification because i sending some dynamic data like "Hi #{user_name}, received #{amount}" and i want to know what exactly was sended.

**Reply by excid3** [reply]
That should really be in the `params` on the `Noticed::Event`. 

You can add columns to `Noticed::Notification` and include it by overriding `recipient_attributes_for` method on the Notifier.

**Reply by phil-6** [reply]
You can also use Virtual attributes on the Noticed Event table and then delegate those values from the Notification to the Event.

For the Migration:

``` ruby
    create_table :noticed_events, id: :uuid do |t|
      t.string :type
      t.belongs_to :record, polymorphic: true, type: :uuid
      t.jsonb :params

      # Custom Fields
      t.string :organisation_id, type: :uuid, as: "((params ->> 'organisation_id')::uuid)", stored: true
      t.virtual :action_type, type: :string, as: "((params ->> 'action_type'))", stored: true
      t.virtual :url, type: :string, as: "((params ->> 'url'))", stored: true

      t.timestamps
    end
```

In the Notification Extensions concern

``` ruby
    delegate :organisation_id, :action_type, :url, to: :event, allow_nil: true
```

I've added this example to the docs in https://github.com/excid3/noticed/pull/389
